---
title: "Assignment 3: investigation of biological effects of SAHA treatment of diabetic human aortic endothelial cells using GSEA and Cytoscape"
output: 
  html_document:
    toc: true
    toc_depth: 2
bibliography: assignment3/A3_michaelapostolides.bib
---

**Name:** Michael Apostolides\

```{r, message=FALSE, echo=FALSE}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")

if (!requireNamespace("gProfileR", quietly = TRUE))
  BiocManager::install("gProfileR")

if (!requireNamespace("gprofiler2", quietly = TRUE))
  BiocManager::install("gprofiler2")

if (!requireNamespace("stringr", quietly = TRUE))
  install.packages("stringr")

if (!requireNamespace("kableExtra", quietly = TRUE))
  install.packages("kableExtra")
```


```{r, message=FALSE, echo=FALSE}
# load needed libraries
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("tidyr")
library("dplyr")
library("ComplexHeatmap")
library(circlize)
library("gprofiler2")
library("stringr")
library(kableExtra)
```


```{r, message=FALSE, echo=FALSE}
#GET THE EXPRESSION DATA
sfiles = getGEOSuppFiles('GSE77108')
fnames = rownames(sfiles)
# there is only one supplemental file
HDAC = read.delim(fnames[1],header=TRUE, check.names = FALSE)
#SUBSET ONLY COLUMNS WHICH have diabetic cells and SAHA or DMSO treatment
SAHA.DMSO <- HDAC[grep("DMSO|SAHA|Geneid", colnames(HDAC))]
diabetic.SAHA.DMSO <- SAHA.DMSO[grep("D-|Geneid", colnames(SAHA.DMSO))]
```


```{r, , message=FALSE, echo=FALSE}
#GENERATE "samples", METADATA ABOUT SAMPLES
samples <- data.frame(lapply(colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)], 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(2)]}))
colnames(samples) <- colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)]
rownames(samples) <- c("treatment")
samples <- data.frame(t(samples), stringsAsFactors=FALSE)

#LABEL CELL TYPES AS "DMSO" TREATED OR "SAHA" TREATED
DMSO <- grep("DMSO", rownames(samples))
SAHA <- grep("SAHA", rownames(samples))
samples$treatment[DMSO] = "DMSO"
samples$treatment[SAHA] = "SAHA"
```


```{r, echo=FALSE, message=FALSE}
#FILTER OUT GENES THAT HAVE LOW COUNTS (CLEANING DATA)
#use edgeR to calculate counts per million (cpms)
cpms = cpm(diabetic.SAHA.DMSO[2:ncol(diabetic.SAHA.DMSO)])
rownames(cpms) <- diabetic.SAHA.DMSO[,1]
#use cpms to determine which of our ENSGs to throw out
keep = rowSums(cpms >1) >=3
diabetic.SAHA.DMSO.filtered = diabetic.SAHA.DMSO[keep,]
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$Geneid
```



```{r map_ensg_to_hgnc, echo=FALSE, message=FALSE, warning=FALSE}
#MAP ENSG TO HGNC SYMBOLS
#get mart object from useast mirror (this mirror was working at the time of running)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")#useast/uswest

conversion_stash <- "SAHA_id_conversion.rds"
if(file.exists(conversion_stash)){
  HDAC_id_conversion <- readRDS(conversion_stash)
} else {
  HDAC_id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = diabetic.SAHA.DMSO.filtered$Geneid,
                               mart = ensembl)
  saveRDS(HDAC_id_conversion, conversion_stash)
}


#rename col in diabetic.SAHA.DMSO.filtered  to match HDAC_id_conversion
colnames(diabetic.SAHA.DMSO.filtered)[colnames(diabetic.SAHA.DMSO.filtered) == "Geneid"] <- "ensembl_gene_id"
#join by ensembl_gene_id
diabetic.SAHA.DMSO.filtered <- dplyr::inner_join(diabetic.SAHA.DMSO.filtered , HDAC_id_conversion, by="ensembl_gene_id" )
#make row names ensembl_gene_id
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$ensembl_gene_id
#remove ensembl_gene_id column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[colnames(diabetic.SAHA.DMSO.filtered) != "ensembl_gene_id"]
#make hgnc_symbol first column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[c(7,1:6)]
#stop("Stop")

```





# Introduction 
The dataset is from [GEO GSE77108](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77108)[@HDAC_2017], and selects one of several treatment analysis in this paper. This paper corresponds to the GEO dataset with ID GSE77108. The treatment selected is SAHA treatment of human aortic endothelial cells obtained from a diabetic individual (dHAECs) line as compared to control DMSO treatment of those same dHAECs. My workflow begins with the RNAseq read counts provided by the authors, which were obtained using the "Counts" feature of the STAR RNAseq aligner[@HDAC_2017].   
Normalization has already been done by the authors using the HPRT1 housekeeping gene[@HDAC_2017]. However, I have gone ahead and normalized the data again anyway using TMM.

```{r TMM-normalization, message=FALSE, echo=FALSE}
#APPLYING TMM NORMALIZATION TO THE DATASET
filtered_data_matrix <- as.matrix(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO)])
rownames(filtered_data_matrix) <- rownames(diabetic.SAHA.DMSO.filtered)
d = DGEList(counts=filtered_data_matrix, group=samples$treatment)
d = calcNormFactors(d)

final.df <- cpm(d)
final.df.log2 <- log2(cpm(d))
# add hgnc symbols as first column
final.df.log2 <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],final.df.log2, stringsAsFactors = FALSE)
colnames(final.df.log2)[1] <- "hgnc_symbol"

# The reason we were getting -Inf values is because the values were "0" prior to log2 transformation
#remove values which have -Inf as values in matrix
final.df <- final.df.log2
final.df[final.df== -Inf] <- 0

```


# Differential Gene Expression 

I used the Benjamini-Hockberg adjustment method. I initially tried the Bonferroni method, but it proved to be too stringent. Some of the pro-inflammantory genes that were deemed "significant" in the [@HDAC_2017] paper were filtered out with Bonferroni, so I used the Benjamini-Hockberg adjustment method instead.\

I used the epiriical Bayes method to compute differential expression, with trend=TRUE set to specify RNAseq data.

```{r create-DE-data-matrix, message=FALSE, echo=FALSE}
#CREATE DATA MATRIX

model_design <- model.matrix(~ samples$treatment)
expressionMatrix <- as.matrix(final.df[,2:ncol(final.df)])
#USE LIMMA PACKAGE
minimalSet <- ExpressionSet(assayData=expressionMatrix)
fit <- lmFit(minimalSet, model_design)
#Apply empircal Bayes to compute differential expression
fit2 <- eBayes(fit,trend=TRUE)

adjust.method = "BH"
#adjust.method="bonferroni"
topfit <- topTable(fit2, 
                   sort.by="none",
                   coef=ncol(model_design),
                   adjust.method = adjust.method,
                   number = nrow(expressionMatrix))

topfit <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],topfit, stringsAsFactors = FALSE)
colnames(topfit)[1] <- "hgnc_symbol"


```

# Non-Thresholded analysis using GSEA

I used GSEA [@GSEA] for non-thresholded analysis. As was discussed in class, I used both the p-value and log-fold change to compute the rank file. I used the same geneset file which was used in the class homework:\    

[Bader lab .gmt file](http://download.baderlab.org/EM_Genesets/February_01_2020/Human/symbol/Human_GOBP_AllPathways_no_GO_iea_February_01_2020_symbol.gmt)

```{r compute-GSEA-ranks}
#compute ranks
ranks <- -log(topfit$adj.P.Val, base=10)*sign(topfit$logFC)
genes <- topfit$hgnc_symbol
rnk.df <- data.frame(GeneName=genes, rank=ranks, stringsAsFactors = FALSE)
#sort table by ranks
rnk.df <- rnk.df[order(rnk.df$rank, decreasing = TRUE),]
rows <- nrow(rnk.df)

#REMOVE ROWS WITH NO HGNC SYMBOL
rnk.df <- rnk.df[rnk.df$GeneName != "",]
rows.no_hgnc <- as.character(rows - nrow(rnk.df))
#print(paste0("Number of entries without an hgnc symbol: ", rows.no_hgnc))

# Write rank file
write.table(rnk.df, file=paste0(getwd(), "/", "sign_hits.unthresholded.rnk"), quote=FALSE, sep='\t', row.names=FALSE)

```


## Summary of enrichment results

I have included a table of the top 15 upregulated and downregulated terms of the GSEA results (Table 1). A summary of the contents for the top hits is as follows:\

* Upregulated: the main themes include: "cognition/learning/memory", "Synaptic signalling", "ion transport/channels", "lipid/lipoprotein" (Table 1).\    
* Downregulated: the main themes include: "E2F", "DNA replication", "ATR signaling/replication stress", "cell cycle checkpoints" (Table 1).

```{r}
neg.file <- paste0(getwd(),"/assignment3/gsea_report_for_na_neg_1585427734453.xls")
neg.file <- read.table(file=neg.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")
pos.file <- paste0(getwd(),"/assignment3/gsea_report_for_na_pos_1585427734453.xls")
pos.file <- read.table(file=pos.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")

top.pos <- head(pos.file$NAME, 15)
top.neg <- head(neg.file$NAME, 15)

df.terms <- data.frame(Upregulated=top.pos, Downregulated=top.neg)
title <- "Table 1: Comparison of top 15 GSEA upregulated and downregulated gene sets"
tab <- kable(df.terms, caption = title)
kable_styling(tab)
```



<!-- How do these results compare to the results from the thresholded analysis in Assignment #2. Compare qualitatively. Is this a straight forward comparison? Why or why not? -->
## Comparison of GSEA and g:profiler results
In the g:profiler results from the previous assignment, the following is seen:\  

* Upregulated: protein localization, lytic vacuole/lysosome, Axonogenesis, wnt signaling, (Ras) GTPase binding, and synapse-related terms.
* Downregulated: DNA replication, cell cycle checkpoint, and chromosome segregation gene sets, retinoblastoma Gene in Cancer 

Below, a comparison is done between the g:profiler results and the GSEA results:\    

* Upregulated comparison: the upregulated terms have only synapse-related terms in common. There are major differences in the top 15 hits
* Downregulated comparison: the downregulated terms have "DNA replication" and "cell cycle checkpoint" terms in common.\    

Comparing the outputs directly using the tables provided is difficult. There are many keywords, and finding themes among terms is time-consuming to do manually.

# Network diagram and HDAC inhibitors

## Initial unannotated network

Below is an initial screenshot of my unannotated network (Figure 1). This network has 3827 edges and 2269 nodes. I used Q-value cuttoffs of 0.1 for the nodes, and 0.375 similarity cuttoff for the edges. 


### Figure 1: Unannotated network of GSEA results. Upregulated (red), Downregulated( blue)\ 
```{r initial-network, out.width= "100%", message=FALSE, echo=FALSE}
knitr::include_graphics(paste0(getwd(), '/assignment3/enrichmentmap-initial.png'))
```

I tried to annotate the regions with many nodes in clusters. However, this proved to be a challenge. There was a large section in the middle which could not be easily read, and was difficult to label meaningfully. I have displayed my attempt below (Figure 2). For more detailed publication-ready figure, I have used AutoAnnotate to obtain clearer clusters (Figure 3)

### Figure 2: Annotated network of GSEA results. Upregulated (red), Downregulated( blue)\ 
```{r annotated-network, out.width= "100%", message=FALSE, echo=FALSE}
knitr::include_graphics(paste0(getwd(), '/assignment3/annotated-EnrichmentMap.png'))

```

## Background on HDACs and HDAC inhibitors

HDACs (Histone deacetylases) are an important class of enzymes involved in epigenetic regulation of a wide variety of genes in the cell [@HDAC_INFO].  When HDACs catalyze deacetylation of key lysine residues on histones, this leads to chromatin condensation, and thus silencing of gene expression [@HDAC_INFO]. The opening of chromatin is done by acetylation of those same lysine residues by HATs (histone acetyltransferases)[@HDAC_INFO].There are also proteins other than histones whose activities are affected by acetylation, such as chromatin remodeling proteins, DNA-binding nuclear receptors, DNA repair enzymes, among others, and changes in the balance of HAT/HDAC activity is known to alter various signaling pathways, including the Wnt pathway [@HDAC_INFO]. 

HDAC inhibitors such as SAHA inhibit the catalytic activity of HDACs, preventing acetyl groups from being removed from target proteins.[@HDAC_CANCER]. SAHA inhibits HDACs by chelating the zinc atom in the catalytic site of HDACs, thus blocking catalytic activity [@SAHA_HDAC].  SAHA is known to inhibit activity of all 11 human class I and class II HDACs [@SAHA_HDAC].

## Common themes between my analysis and the Raferi et. al. paper's analysis

In Assignment 2, I showed that regulation of 6 EP300 target genes (VCAN, CD58, CD200,  SIGIRR, CCL2, ICAM2) were downregulated in SAHA-treated dHAECs as compared to DMSO-treated dHAECs. The [@HDAC_2017] paper did not specifically find gene sets in their GSEA analysis which corresponded to these genes, and there are no gene sets in the .gmt file I used for this assignment that relate specifically to EP300. Therefore, I will compare my GSEA+Enrichment Map results with those from the [@HDAC_2017] paper

In the [@HDAC_2017] paper, they also used GSEA and Cytoscape/Enrichment Map to visualize the themes among gene sets. There are many commonalities between Figure 3 (my network) and theirs (Figure 3 in [@HDAC_2017] ). Cell cycle and DNA replication, RNA processing and transport, Insulin siglnaling, and neuronal systems/ion channels/synapses are among the pathways in their figure which are common to Figure 3 in my analysis. 


## Interpretation of themes in network diagram

Since HDACs are histone modifying proteins, downregulation of the theme "histone modification" is consistent with SAHA treatment (Figure 3). 

Class I HDACs, when over-expressed, result in over-proliferation phenotypes, and their inhibition is known to cause cell cycle arrest, growth inhibition, and apoptosos in cancer [@HDAC_CANCER]. In Figure 3, we see themes such as "cell cycle g2", "chromatid metaphase segregation", and "mitotic checkpoint integrity" being downregulated, which are related to cell cycle arrest and growth inhibition. This is consistent with inhibition of class I HDACs by SAHA. 

Class II HDACs have a more mixed function, and can be pro-proliferative or tumor suppressors depending on the context. HDAC4 activity is known to promote DNA repair and result in a block in cell proliferation [@HDAC_CANCER], and HDAC4 activity is known to be inhibited by SAHA, since it is a class II HDAC [@SAHA_HDAC]. In Figure 3, the downregulation of themes "DNA repair/replication" (the theme with the most gene sets), "double-stranded break repair", "D-loop recombination repair" are consistent with inhibition of HDAC4 by SAHA. 

HDAC4 is also believed to positively regulate SUMOylation of proteins involved in cognitive function [@HDAC4_SUMO]. Since SAHA downregulates HDAC4, this could explain the downregulation of the "SUMOylation, RNA nuclear export" theme in Figure 3.

The wide range of proteins whose functions are regulated by HDAC-dependent deacetylation is reflected in the diverse themes present in Figure 3. It is not possible to investigate all of them in detail. Overall, the literature supports the variety of themes in Figure 3, which result from HDAC inhibition by SAHA in dHAECs. 

### Figure 3: Themed network diagram of GSEA results. Upregulated (red), Downregulated( blue)\    
```{r themed-network-diagram, out.width= "100%",  message=FALSE, echo=FALSE}
knitr::include_graphics(paste0(getwd(), '/assignment3/annotated.dHAECs.SAHA-v2.png'))
```


# Wnt pathway - further investigation


## Background and rationale

I have chosen the Wnt pathway as a pathway to investigate in the GSEA non-thresholded results. Wnt-related gene sets are make up 11/296 significant gene sets with p-value threshold of 0.01, including ranks #23 (NEGATIVE REGULATION OF CANONICAL WNT SIGNALING PATHWAY%GOBP%GO:0090090) and #44 ("NEGATIVE REGULATION OF WNT SIGNALING PATHWAY%GOBP%GO:0030178") in "na_pos" GSEA results. There are no Wnt-related terms in the "na_neg" downregulated sets, suggesting that upregulation of Wnt-related genes is uniquely seen in SAHA-treated dHAECs. 

The Wnt pathway is an evolutionarily conserved signalling pathway, and its aberrant activation is know to play a central role in a wide variety of cancers [@WNT_CANCER]. 
There is also literature support for modulation of the Wnt pathway by SAHA.  In the (GÃ¶tze et al. 2014) paper, they show that SAHA attenuates proliferation and induces apoptosis in Wnt-activated colorectal carcinoma cell lines. In the (Sujeong Jang 2018b) paper, they show that expression of most of the Wnt-related genes they were investigating were highly increased following treatment with HDAC inhibitors.

The Wnt pathway is a well-defined pathway with literature support for SAHA effects, so it makes a good candidate for more detailed investigation. 


```{r}
#THIS SECTION GETS INTERSECTION OF WIKIPATHWAY GENES AND LEADING EDGE GENES FROM GO:0090090

pval <- 0.01
pos.file.sign <- pos.file[pos.file$NOM.p.val < pval,]
wnt.pos.sign <- pos.file.sign[grep("WNT", pos.file.sign$NAME),]

neg.file.sign <- neg.file[neg.file$NOM.p.val < pval,]
wnt.neg.sign <- neg.file.sign[grep("WNT", neg.file.sign$NAME),]

leading.edge.genes.file <- paste0(getwd(), "/assignment3/GSEA_outdir/A3_GSEA_UNTHRESHOLDD.GseaPreranked.1585427734453/NEGATIVE\ REGULATION\ OF\ CANONICAL\ WNT\ SIGNALING\ PATHWAY%GOBP%GO:0090090.xls")
leading.edge.genes.file <- read.table(file=leading.edge.genes.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")
leading.edge <- leading.edge.genes.file[leading.edge.genes.file$CORE.ENRICHMENT == "Yes",]
leading.edge <- data.frame(PROBE=leading.edge$PROBE, RANK.METRIC.SCORE=leading.edge$RANK.METRIC.SCORE)
leading.edge.genes <- leading.edge$PROB

wikipathway.genes <- paste0(getwd(),  "/assignment3/Wnt.pathway.WikiPathways.txt")
wikipathway.genes <- read.table(file=wikipathway.genes, header = F, stringsAsFactors = F, sep="\t", quote = "\"")
wikipathway.genes <- wikipathway.genes$V1
intersected.genes <- data.frame(genes=noquote(intersect(wikipathway.genes, leading.edge.genes)))

title <- "Panel 1: Genes common between GO term GO0090090 and WP428 "
tab <- kable(intersected.genes, caption = title)
kable_styling(tab)
```


## Annotated Wnt pathway diagram from WikiPathways

I have downloaded the WP428 (Wnt Signaling) WikiPathways network using the "Import Network from Public Databases" feature of Cytoscape, and annotated it with the rank file I used as input to my GSEA run (Figure 4). Additionally, the GO term "GO:0090090", "negative regulation of canonical Wnt signaling pathway", which is hit #23 of GSEA's "na_pos" results, has 8 genes in common with WikiPathway WP428 (Panel 1). All of these genes are in the leading edge of this GSEA result.
There are 44 other genes in the leading edge which are not listed in Figure 4, but which are known to negatively regulated Wnt signalling.

DKK1 and SFRP1 are both highly ranked Wnt pathway inhibitors, as can be seen by the (----| ) inhibitory arrows in Figure 4. SFRP1 inhibits signaling of Wnt proteins directly, whereas  DKK1 inhibits LRP5/6, which in turn interact with Frizzled(FZD) receptors (Figure 4). DKK1, a known Wnt antagonist, is epigenetically silenced in a medulloblastoma cell line, and is upregulated upon HDAC inhibition [@DKK1_HDAC]. Epigenetic silencing of SFRP1 leads to increased survival of renal and breast cancers, and re-rexpression of SFRP1 upon treatment with HDAC inhibitors coincides with reduced cell growth and apoptosis in chemoresistant cancers [@SFRP1_HDAC]. Additionally, promoter methylation (i.e. repression of transcription) of Wnt antagonists DKK1 and SFRP1 are known events which lead to hyper-activation of the Wnt pathway [@WNT_CANCER]. It then follows that an increase in their expression would have the effect of suppressing Wnt signalling. 



### Figure 4: Highlighting of leading edge genes in Wnt pathway (Wikipathways, WP428). Upregulated (red), Downregulated( blue). Circled genes are leading edge genes of GSEA GO term "GO 0090090"          

```{r out.width= "100%", message=FALSE, echo=FALSE}
knitr::include_graphics(paste0(getwd(), '/assignment3/Wnt-WikiPathways-Annotated.trimmed.png'))
```

## Wnt signalling and diabetes

In the original [@HDAC_2017] paper from which this dataset was taken, they propose SAHA as a potential treatment for diabetes on the basis that SAHA causes suppression of EP300 target genes, and has the potential to reduce inflammation. However, there is no mention of the Wnt signalling pathway in this paper.

A search of the literature finds that over-activation of the Wnt signalling pathway may be a contributing factor to diabetes. TCF7L2, which is upregulated in islets of type II diabetes patients [@WNT_ACTIVATION_DIABETES] is inhibited by SOX17 (Figure 4) through inhibition of upstream singalling. This same paper shows that B-catenin, one of the downstream effects of the Wnt pathway, is not detectable in healthy human beta-cells, but is strongly upregulated in type II diabetes [@WNT_ACTIVATION_DIABETES]. Additionally, terminal effectors of Wnt signalling are shown to be upregulated in type II diabetes [@WNT_ACTIVATION_DIABETES]. Another study suggests that inhibition of Wnt signalling may be beneficial in metabolic disease [@WNT_INHIBIT_DIABETES]. 

I could not find anything in the literature that specifically investigated the effects of HDAC-inhibitors on the Wnt pathway in diabetes. It is possible that no studies have yet been done which specifically investigate this mechanism. The results in Figure 4 suggest that investigation of SAHA as a treatment option for Wnt-activated diabetes may be a worthwhile research pursuit.

# Summary

Visualization of GSEA results of SAHA-treated dHAECs in Cytoscape demonstrates that the HDAC inhibitor SAHA affects a huge number of biological processes. A thorough investigation of all these processes is beyond the scope of this assignment. My investigation of Wnt signalling pathway alterations in the [@HDAC_2017] dataset suggests the authors cherry-picked EP300 target genes as an effect of treating dHAECs with SAHA, and that other diabetes-related effects are seen.  Further investigation into the effects of SAHA on  cellular processes is needed. Overall, my findings are consistent with the literature, with the results and conclusions of the [@HDAC_2017]  paper, and opens up the possibility for extending SAHA's scope for the treatment of diabetes.


# R package references
GEOmetadb https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html \ 

edgeR https://bioconductor.org/packages/release/bioc/html/edgeR.html \ 

biomaRt https://bioconductor.org/packages/release/bioc/html/biomaRt.html \ 

tidyr https://cran.r-project.org/web/packages/tidyr/index.html \ 

dplyr https://cran.r-project.org/web/packages/dplyr/index.html \ 

ComplexHeatmap https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html \ 

circlize https://cran.r-project.org/web/packages/circlize/index.html \ 

gprofiler2 https://cran.r-project.org/web/packages/gprofiler2/index.html\

stringr https://cran.r-project.org/web/packages/stringr/stringr.pdf\

kableExtra https://cran.r-project.org/web/packages/kableExtra/index.html\

# Paper references




