---
title: "Assignment 1: Data preparation of expression data for SAHA treatment of diabetic endothelial cells"
output: html_notebook
---

**What are the control and test conditions of the dataset?**  
Control condition: Diabetic endothelial cells treated with DMSO  
Test condition: Diabetic endothelial cells  treated with the HDAC inhibitor SAHA  

**Why is the dataset of interest to you?**  
I have worked with cancer data sets and studied cancers for many years. I have heard of HDAC inhibitors and knew that they were used as anti-cancer drugs, and was interested to explore their role in being a possible treatment for diabetes  


```{r}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")
```



```{r}
# load needed libraries
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("tidyr")
library("dplyr")

```


```{r}
#GET THE EXPRESSION DATA
sfiles = getGEOSuppFiles('GSE77108')
fnames = rownames(sfiles)
# there is only one supplemental file
HDAC = read.delim(fnames[1],header=TRUE, check.names = FALSE)
#SUBSET ONLY COLUMNS WHICH have diabetic cells and SAHA or DMSO treatment
SAHA.DMSO <- HDAC[grep("DMSO|SAHA|Geneid", colnames(HDAC))]
diabetic.SAHA.DMSO <- SAHA.DMSO[grep("D-|Geneid", colnames(SAHA.DMSO))]
```


```{r}
#GENERATE "samples", METADATA ABOUT SAMPLES
samples <- data.frame(lapply(colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)], 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(2)]}))
colnames(samples) <- colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)]
rownames(samples) <- c("treatment")
samples <- data.frame(t(samples), stringsAsFactors=FALSE)

#LABEL CELL TYPES AS "DMSO" TREATED OR "SAHA" TREATED
DMSO <- grep("DMSO", rownames(samples))
SAHA <- grep("SAHA", rownames(samples))
samples$treatment[DMSO] = "DMSO"
samples$treatment[SAHA] = "SAHA"
```


```{r}
#FILTER OUT GENES THAT HAVE LOW COUNTS (CLEANING DATA)
#use edgeR to calculate counts per million (cpms)
cpms = cpm(diabetic.SAHA.DMSO[2:ncol(diabetic.SAHA.DMSO)])
rownames(cpms) <- diabetic.SAHA.DMSO[,1]
#use cpms to determine which of our ENSGs to throw out
keep = rowSums(cpms >1) >=3
diabetic.SAHA.DMSO.filtered = diabetic.SAHA.DMSO[keep,]
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$Geneid
```


**Were there expression values that could not be mapped to current HUGO symbols?**  
There were ENSG symbols which could not be mapped to HGNC symbols using biomaRt. I removed these from the analysis, since we are required to use HGNC symbols as rownames.  
Initially, my filtered data frame had 13648 rows. My final data frame has 12609 rows. This difference of 1039 corresponds to both ENSG IDs that biomart did not recognize (7000), as well as ENSG IDs which returned no HGNC symbol (339).  



```{r}
#MAP ENSG TO HGNC SYMBOLS
#get mart object from useast mirror (this mirror was working at the time of running)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")

conversion_stash <- "SAHA_id_conversion.rds"
if(file.exists(conversion_stash)){
  HDAC_id_conversion <- readRDS(conversion_stash)
} else {
  HDAC_id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = diabetic.SAHA.DMSO.filtered$Geneid,
                               mart = ensembl)
  saveRDS(HDAC_id_conversion, conversion_stash)
}

#NUMBER OF ROWS IN diabetic.SAHA.DMSO.filtered BEFORE REMOVING
print("NUMBER OF ROWS IN diabetic.SAHA.DMSO.filtered BEFORE REMOVING"); nrow(diabetic.SAHA.DMSO.filtered)
#REMOVE ROWS WHERE HGNC SYMBOLS ARE EMPTY STRING
HDAC_id_conversion <- HDAC_id_conversion[HDAC_id_conversion$hgnc_symbol != "",]

#NUMBER OF ENSG WITH HGNC SYMBOLS
ENSG.with.HGNC.symbols <- HDAC_id_conversion$ensembl_gene_id
print("ENSG.with.HGNC.symbols:"); length(ENSG.with.HGNC.symbols)

#NUMBER OF ENSGs WHICH DIDN'T RETURN HGNC SYMBOLS
print("NUMBER OF ENSGs WHICH DIDN'T RETURN HGNC SYMBOLS:"); nrow(diabetic.SAHA.DMSO.filtered) - length(ENSG.with.HGNC.symbols)


#rename col in diabetic.SAHA.DMSO.filtered  to match HDAC_id_conversion
colnames(diabetic.SAHA.DMSO.filtered)[colnames(diabetic.SAHA.DMSO.filtered) == "Geneid"] <- "ensembl_gene_id"
#join by ensembl_gene_id
diabetic.SAHA.DMSO.filtered <- dplyr::inner_join(diabetic.SAHA.DMSO.filtered , HDAC_id_conversion, by="ensembl_gene_id" )
#make row names hgnc_symbol
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$hgnc_symbol
#remove hgnc_symbol column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[colnames(diabetic.SAHA.DMSO.filtered) != "hgnc_symbol"]

#concatenating ensgid and hgnc
#tmp <- HDAC_id_conversion %>% unite(tmp, c("ensembl_gene_id","hgnc_symbol")) 

#HPRT1
#HPRT1_ensg <- HDAC_id_conversion$ensembl_gene_id[grep("HPRT1", HDAC_id_conversion$hgnc_symbol)]
#diabetic.SAHA.DMSO.filtered[HPRT1_ensg,]
```



```{r}
#APPLYING TMM TO THE DATASET
filtered_data_matrix <- as.matrix(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO)])
rownames(filtered_data_matrix) <- rownames(diabetic.SAHA.DMSO.filtered)
d = DGEList(counts=filtered_data_matrix, group=samples$treatment)
d = calcNormFactors(d)
normalized_counts <- cpm(d)

```


**MAKE MDS PLOT TO SEE THE DISTANCE BETWEEN THE SAMPLES**  
DMSO and SAHA treated diabetic cells seem to cluster together for the mostpart, with the exception of D-SAHA-2 along dimension 2. 
```{r}

plotMDS(d, labels=rownames(samples),
  col = c("darkgreen","blue")[factor(samples$treatment)])
```



```{r}
#PLOT NON-NORMALIZED DATA
#Distribution of our data - Density plot
counts_density <- apply(log2(cpm(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO.filtered)])), 2, density)
#calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))
#plot the first density plot to initialize the plot
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
     ylab="Smoothing density of log2-CPM", main="Density curve for non-normalized counts per million", cex.lab = 0.85)
#plot each line
for (i in 1:length(counts_density)) lines(counts_density[[i]], col=cols[i], lty=ltys[i])
#create legend
legend("topright", rownames(samples),  
       col=cols, lty=ltys, cex=0.75, 
       border ="blue",  text.col = "green4", 
       merge = TRUE, bg = "gray90") 

#PLOT NORMALIZED DATA

counts_density <- apply(log2(cpm(d)), 2, density)
#calculate the limits across all the samples
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x)); 
  ylim <- range(c(ylim, counts_density[[i]]$y))
}
cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))
#plot the first density plot to initialize the plot
plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n", 
     ylab="Smoothing density of log2-CPM", main="Density curve for TMM normalized counts per million", cex.lab = 0.85)
#plot each line
for (i in 1:length(counts_density)) lines(counts_density[[i]], col=cols[i], lty=ltys[i])
#create legend
legend("topright", rownames(samples),  
       col=cols, lty=ltys, cex=0.75, 
       border ="blue",  text.col = "green4", 
       merge = TRUE, bg = "gray90") 

#identical(cpm(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO.filtered)]), cpm(d$counts))
```

NORMALIZATION

-In the paper, it is stated that" ...(aligned using) STAR. The tool feature Counts was used to generate a matrix of read counts per gene and the edgeR software was then used to determine differential gene expression", and also "Gene expression changes were determined by normalizing against HPRT1." [1]  
-It seems like normalization has already been done on these samples, since upon TMM normalization of the data, there is only a very minimal shift in density curves  


```{r}
#estimate common and tagwise dispersion, model_design not used since there is only 1 dimension (i.e. cell type)
d <- estimateDisp(d)
plotBCV(d,col.tagwise = "black",col.common = "red")
```

Final data frame "final.df ".
As specified in the assignment, this data frame has x=6 numeric columns, one for each sample, and the rownames are HGNC symbols obtained from biomaRt. ENSG IDs are not included in the data frame, since this would result in 1 non-numeric column.  Rows which did not have a corresponding HGNC symbol were removed, since unique HGNC rownames could not be given if they are included.
```{r}
final.df <- log2(cpm(d))

```

```{r}
#compile R notebook
rmarkdown::render('assignment1.Rmd', output_format = 'html_document')
```

# Remaining interpretation questions

**How did you handle replicates?**  
I have 6 samples, and split them into 2 groups: 3 replicates for DMSO treatment and 3 replicates for SAHA treatment.

**Were there expression values that were not unique for specific genes? How did you handle these?**  
No, there were not duplicate HGNC identifiers.

**How many outliers were removed?**  
 I did not remove outliers


**What is the final coverage of your dataset?**  
The final coverage is 12609 HGNC rows in my final data frame.


## References

[1] Rafehi H, Kaspi A, Ziemann M, Okabe J, Karagiannis TC, El-Osta A. "Systems approach to the pharmacological actions of HDAC inhibitors reveals EP300 activities and convergent mechanisms of regulation in diabetes." Epigenetics. 2017;12(11):991-1003. doi: 10.1080/15592294.2017.1371892.

