---
title: "Assignment 2: Expression analysis for SAHA treatment of diabetic endothelial cells"
output: 
  html_document:
    toc: true
    toc_depth: 2
bibliography: assignment2.bib
---



# Install and load needed packages
```{r, message=FALSE, echo=FALSE}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")


```


```{r, message=FALSE, echo=FALSE}
# load needed libraries
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("tidyr")
library("dplyr")
library("ComplexHeatmap")
library(circlize)

```


# Download expression data from GEO
```{r, message=FALSE, echo=FALSE}
#GET THE EXPRESSION DATA
sfiles = getGEOSuppFiles('GSE77108')
fnames = rownames(sfiles)
# there is only one supplemental file
HDAC = read.delim(fnames[1],header=TRUE, check.names = FALSE)
#SUBSET ONLY COLUMNS WHICH have diabetic cells and SAHA or DMSO treatment
SAHA.DMSO <- HDAC[grep("DMSO|SAHA|Geneid", colnames(HDAC))]
diabetic.SAHA.DMSO <- SAHA.DMSO[grep("D-|Geneid", colnames(SAHA.DMSO))]
```


```{r, echo=FALSE}
#GENERATE "samples", METADATA ABOUT SAMPLES
samples <- data.frame(lapply(colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)], 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(2)]}))
colnames(samples) <- colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)]
rownames(samples) <- c("treatment")
samples <- data.frame(t(samples), stringsAsFactors=FALSE)

#LABEL CELL TYPES AS "DMSO" TREATED OR "SAHA" TREATED
DMSO <- grep("DMSO", rownames(samples))
SAHA <- grep("SAHA", rownames(samples))
samples$treatment[DMSO] = "DMSO"
samples$treatment[SAHA] = "SAHA"
```


```{r, echo=FALSE}
#FILTER OUT GENES THAT HAVE LOW COUNTS (CLEANING DATA)
#use edgeR to calculate counts per million (cpms)
cpms = cpm(diabetic.SAHA.DMSO[2:ncol(diabetic.SAHA.DMSO)])
rownames(cpms) <- diabetic.SAHA.DMSO[,1]
#use cpms to determine which of our ENSGs to throw out
keep = rowSums(cpms >1) >=3
diabetic.SAHA.DMSO.filtered = diabetic.SAHA.DMSO[keep,]
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$Geneid
```



```{r, echo=FALSE}
#MAP ENSG TO HGNC SYMBOLS
#get mart object from useast mirror (this mirror was working at the time of running)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")

conversion_stash <- "SAHA_id_conversion.rds"
if(file.exists(conversion_stash)){
  HDAC_id_conversion <- readRDS(conversion_stash)
} else {
  HDAC_id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = diabetic.SAHA.DMSO.filtered$Geneid,
                               mart = ensembl)
  saveRDS(HDAC_id_conversion, conversion_stash)
}


#rename col in diabetic.SAHA.DMSO.filtered  to match HDAC_id_conversion
colnames(diabetic.SAHA.DMSO.filtered)[colnames(diabetic.SAHA.DMSO.filtered) == "Geneid"] <- "ensembl_gene_id"
#join by ensembl_gene_id
diabetic.SAHA.DMSO.filtered <- dplyr::inner_join(diabetic.SAHA.DMSO.filtered , HDAC_id_conversion, by="ensembl_gene_id" )
#make row names ensembl_gene_id
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$ensembl_gene_id
#remove ensembl_gene_id column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[colnames(diabetic.SAHA.DMSO.filtered) != "ensembl_gene_id"]
#make hgnc_symbol first column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[c(7,1:6)]
#stop("Stop")

```





# INTRODUCTION  
The dataset is from [GEO GSE77108](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77108)[@HDAC_2017], and selects one of several treatment analysis in this paper. This paper corresponds to the GEO dataset with ID GSE77108. The treatment selected is SAHA treatment of a diabetic endothelial cell line as compared to control DMSO treatment. My workflow begins with the RNAseq read counts provided by the authors, which were obtained using the "Counts" feature of the STAR RNAseq aligner[@HDAC_2017].   
Normalization has already been done by the authors using the HPRT1 housekeeping gene[@HDAC_2017]. However, I have gone ahead and normalized the data again anyway using TMM:

## TMM normalization
```{r}
#APPLYING TMM NORMALIZATION TO THE DATASET
filtered_data_matrix <- as.matrix(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO)])
rownames(filtered_data_matrix) <- rownames(diabetic.SAHA.DMSO.filtered)
d = DGEList(counts=filtered_data_matrix, group=samples$treatment)
d = calcNormFactors(d)

final.df <- cpm(d)
final.df.log2 <- log2(cpm(d))
# add hgnc symbols as first column
final.df.log2 <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],final.df.log2, stringsAsFactors = FALSE)
colnames(final.df.log2)[1] <- "hgnc_symbol"




# The reason we were getting -Inf values is because the values were "0" prior to log2 transformation
#neg.inf.ensg_ids <- rownames(df[apply(df, 1, function(r) any(r %in% c(-Inf))),])
#look.at <- final.df[neg.inf.ensg_ids,]


#remove values which have -Inf as values in matrix
#final.df <- final.df.log2[!apply(final.df.log2, 2, function(r) any(r %in% c(-Inf))),]
final.df <- final.df.log2
final.df[final.df== -Inf] <- 0
#final.df <- do.call(data.frame,lapply(final.df, function(x) replace(x, is.infinite(x),0)))

#This one still has -Inf??
#df.MAST1 <- final.df[final.df$hgnc_symbol == "MAST1",]
#df.MAST1[df.MAST1== -Inf] <- 0
#do.call(data.frame,lapply(df.MAST1, function(x) replace(x, is.infinite(x),0)))

```





The author's normalization is evident from the fact that there is only a very minimal shift in the density curves of the non-TMM-normalized (diabetic.SAHA.DMSO.filtered) and TMM-normalized(d) data below:



# Differential Gene Expression 


**MAKE MDS PLOT TO SEE THE DISTANCE BETWEEN THE SAMPLES**  
DMSO and SAHA treated diabetic cells seem to cluster together for the mostpart, with the exception of D-SAHA-2 along dimension 2. 
```{r}

# plotMDS(d, labels=rownames(samples),
#   col = c("darkgreen","blue")[factor(samples$treatment)])
```



```{r}
#compile R notebook
#rmarkdown::render('assignment2.Rmd', output_format = 'html_document')
```


The heatmap shows that DMSO and SAHA treated are clustering separately.


```{r}
#Make HEATMAP ==> TAKES LONG TIME, ONLY PRINT HEATMAP AT VERY END
# heatmap_matrix <- head(final.df, 1000)
# 
# heatmap_matrix <- t(scale(t(heatmap_matrix)))
# heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, max(heatmap_matrix)), c("blue", "white", "red"))
# 
# current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
#                                show_row_dend = TRUE,
#                                show_column_dend = TRUE, 
#                                col=heatmap_col,
#                                show_column_names = TRUE, 
#                                show_row_names = FALSE,
#                                show_heatmap_legend = TRUE
#                                )
#current_heatmap

```


```{r}
#CREATE DATA MATRIX

model_design <- model.matrix(~ samples$treatment)
expressionMatrix <- as.matrix(final.df[,2:ncol(final.df)])
#USE LIMMA PACKAGE
minimalSet <- ExpressionSet(assayData=expressionMatrix)
fit <- lmFit(minimalSet, model_design)
#Apply empircal Bayes to compute differential expression
fit2 <- eBayes(fit,trend=TRUE)

adjust.method = "BH"
#adjust.method="bonferroni"
topfit <- topTable(fit2, 
                   sort.by="none",
                   coef=ncol(model_design),
                   adjust.method = adjust.method,
                   number = nrow(expressionMatrix))

# output_hits <- merge(final.df[,1],
#                      topfit,
#                      by.y=0,by.x=1,
#                      all.y=TRUE)
topfit <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],topfit, stringsAsFactors = FALSE)
colnames(topfit)[1] <- "hgnc_symbol"

# SORT AND FILTER BY PVAL
pval <- 0.05
#ORDER BY P-VALUE
output_hits <- topfit[order(topfit$P.Value),]
length(which(output_hits$adj.P.Val < pval))
#FILTER BY P-VALUE THRESHOLD
output_hits.pval <-  output_hits[which(output_hits$adj.P.Val < pval),]
# SORT BY DECREASING EXPRESSION VALUE
output_hits.pval <- output_hits[order(output_hits.pval$logFC, decreasing = TRUE),]

testing <- 0
if (testing){
    output_hits.pval[order(output_hits.pval$AveExpr),]
    nrow(output_hits.pval)
    # MAYBE COLOUR THESE DOTS IN THE MA PLOT (THE INCREASE/DECREASE EXPRESSION PLOT)
    output_hits.pval[output_hits.pval$hgnc_symbol == "MAST1",]
    output_hits.pval[output_hits.pval$hgnc_symbol == "VCAN",]
    output_hits.pval[output_hits.pval$hgnc_symbol == "CD58",]
    output_hits.pval[output_hits.pval$hgnc_symbol == "CD200",]
    output_hits.pval[output_hits.pval$hgnc_symbol == "SIGIRR",]
}
#UPREGULATED
output_hits.pval.up <- output_hits.pval[which(output_hits.pval$logFC > 0),]
length(which(output_hits.pval.up$adj.P.Val < pval))
up.down="up."
#cat(noquote(output_hits.pval$hgnc_symbol[which(output_hits.pval$adj.P.Val < pval)]))
#write(output_hits.pval$hgnc_symbol[which(output_hits.pval.up$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )

# #TEST
# grep("NOTCH3", output_hits.pval.up)
# 
#DOWNREGULATED

output_hits.pval.down <- output_hits.pval[which(output_hits.pval$logFC < 0),]
output_hits.pval.down <- output_hits.pval[order(output_hits.pval$logFC, decreasing = FALSE),]
length(which(output_hits.pval.up$adj.P.Val < pval))
up.down="down."
# write(output_hits.pval$hgnc_symbol[which(output_hits.pval.down$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )
```


```{r}
# http://web.mit.edu/~r/current/arch/i386_linux26/lib/R/library/geneplotter/html/plotMA.html

#plot <- limma::plotMA(output_hits[c(3,2,6)])
# plot <- limma::plotMA(topfit)
#can take average expression values of samples for each condition, enter only 2 columns
# expr.values <- final.df[,2:ncol(final.df)]
# expr.values <- cbind(rowMeans(expr.values[,c(1:3)]), rowMeans(expr.values[,c(4:6)]))
# colnames(expr.values) <- c("DMSO", "SAHA")
# plot(expr.values)

#USE INDEX VALUES OF output_hits.up and output_hits.down to 
plot(topfit$AveExpr,topfit$logFC)
#plot(output_hits$AveExpr,output_hits$logFC)
pval <- 0.05
down.index <- which((topfit$adj.P.Val < pval) & (topfit$logFC < 0))
up.index <- which((topfit$adj.P.Val < pval) & (topfit$logFC > 0))
# output_hits.pval <-  output_hits[which(output_hits$adj.P.Val < pval),]
#up.index <-
points(topfit$AveExpr[up.index], topfit$logFC[up.index], col = "green")
points(topfit$AveExpr[down.index], topfit$logFC[down.index], col = "red")
# points(up) # can label individual points, 
# points(down)
#plotMA(log2(ca125_exp[,c(3,4)]), ylab="M - ratio log expression", main="CA125 + vs - - example")
```


# References
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4670218/ # shows CPA4 upregulated upon HDAC treatment
https://www.ncbi.nlm.nih.gov/pubmed/25050608 # shows HDAC inhibitors attenuate Wnt signalling pathway (this is found in gene set bonferroni.0.005.gene_hits.txt, BH.5e-06.gene_hits.txt)

