---
title: "Assignment 2: Differential expression analysis and investigation of biological effects of SAHA treatment of diabetic human aortic endothelial cells"
output: 
  html_document:
    toc: true
    toc_depth: 2
bibliography: A2_michaelapostolides.bib
---

**Name:** Michael Apostolides\

```{r, message=FALSE, echo=FALSE}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")

if (!requireNamespace("gProfileR", quietly = TRUE))
  BiocManager::install("gProfileR")

if (!requireNamespace("gprofiler2", quietly = TRUE))
  BiocManager::install("gprofiler2")

if (!requireNamespace("stringr", quietly = TRUE))
  install.packages("stringr")

if (!requireNamespace("kableExtra", quietly = TRUE))
  install.packages("kableExtra")
```


```{r, message=FALSE, echo=FALSE}
# load needed libraries
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("tidyr")
library("dplyr")
library("ComplexHeatmap")
library(circlize)
library("gprofiler2")
library("stringr")
library(kableExtra)
```


# Download expression data from GEO
```{r, message=FALSE, echo=FALSE}
#GET THE EXPRESSION DATA
sfiles = getGEOSuppFiles('GSE77108')
fnames = rownames(sfiles)
# there is only one supplemental file
HDAC = read.delim(fnames[1],header=TRUE, check.names = FALSE)
#SUBSET ONLY COLUMNS WHICH have diabetic cells and SAHA or DMSO treatment
SAHA.DMSO <- HDAC[grep("DMSO|SAHA|Geneid", colnames(HDAC))]
diabetic.SAHA.DMSO <- SAHA.DMSO[grep("D-|Geneid", colnames(SAHA.DMSO))]
```


```{r, echo=FALSE}
#GENERATE "samples", METADATA ABOUT SAMPLES
samples <- data.frame(lapply(colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)], 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(2)]}))
colnames(samples) <- colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)]
rownames(samples) <- c("treatment")
samples <- data.frame(t(samples), stringsAsFactors=FALSE)

#LABEL CELL TYPES AS "DMSO" TREATED OR "SAHA" TREATED
DMSO <- grep("DMSO", rownames(samples))
SAHA <- grep("SAHA", rownames(samples))
samples$treatment[DMSO] = "DMSO"
samples$treatment[SAHA] = "SAHA"
```


```{r, echo=FALSE, message=FALSE}
#FILTER OUT GENES THAT HAVE LOW COUNTS (CLEANING DATA)
#use edgeR to calculate counts per million (cpms)
cpms = cpm(diabetic.SAHA.DMSO[2:ncol(diabetic.SAHA.DMSO)])
rownames(cpms) <- diabetic.SAHA.DMSO[,1]
#use cpms to determine which of our ENSGs to throw out
keep = rowSums(cpms >1) >=3
diabetic.SAHA.DMSO.filtered = diabetic.SAHA.DMSO[keep,]
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$Geneid
```



```{r map_ensg_to_hgnc, echo=FALSE, message=FALSE, warning=FALSE}
#MAP ENSG TO HGNC SYMBOLS
#get mart object from useast mirror (this mirror was working at the time of running)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "uswest")#useast

conversion_stash <- "SAHA_id_conversion.rds"
if(file.exists(conversion_stash)){
  HDAC_id_conversion <- readRDS(conversion_stash)
} else {
  HDAC_id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = diabetic.SAHA.DMSO.filtered$Geneid,
                               mart = ensembl)
  saveRDS(HDAC_id_conversion, conversion_stash)
}


#rename col in diabetic.SAHA.DMSO.filtered  to match HDAC_id_conversion
colnames(diabetic.SAHA.DMSO.filtered)[colnames(diabetic.SAHA.DMSO.filtered) == "Geneid"] <- "ensembl_gene_id"
#join by ensembl_gene_id
diabetic.SAHA.DMSO.filtered <- dplyr::inner_join(diabetic.SAHA.DMSO.filtered , HDAC_id_conversion, by="ensembl_gene_id" )
#make row names ensembl_gene_id
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$ensembl_gene_id
#remove ensembl_gene_id column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[colnames(diabetic.SAHA.DMSO.filtered) != "ensembl_gene_id"]
#make hgnc_symbol first column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[c(7,1:6)]
#stop("Stop")

```





# INTRODUCTION  
The dataset is from [GEO GSE77108](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77108)[@HDAC_2017], and selects one of several treatment analysis in this paper. This paper corresponds to the GEO dataset with ID GSE77108. The treatment selected is SAHA treatment of human aortic endothelial cells obtained from a diabetic individual (dHAECs) line as compared to control DMSO treatment of those same dHAECs. My workflow begins with the RNAseq read counts provided by the authors, which were obtained using the "Counts" feature of the STAR RNAseq aligner[@HDAC_2017].   
Normalization has already been done by the authors using the HPRT1 housekeeping gene[@HDAC_2017]. However, I have gone ahead and normalized the data again anyway using TMM:

## TMM normalization
```{r TMM-normalization}
#APPLYING TMM NORMALIZATION TO THE DATASET
filtered_data_matrix <- as.matrix(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO)])
rownames(filtered_data_matrix) <- rownames(diabetic.SAHA.DMSO.filtered)
d = DGEList(counts=filtered_data_matrix, group=samples$treatment)
d = calcNormFactors(d)

final.df <- cpm(d)
final.df.log2 <- log2(cpm(d))
# add hgnc symbols as first column
final.df.log2 <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],final.df.log2, stringsAsFactors = FALSE)
colnames(final.df.log2)[1] <- "hgnc_symbol"

# The reason we were getting -Inf values is because the values were "0" prior to log2 transformation
#remove values which have -Inf as values in matrix
final.df <- final.df.log2
final.df[final.df== -Inf] <- 0

```


# Differential Gene Expression 

I used the Benjamini-Hockberg adjustment method. I initially tried the Bonferroni method, but it proved to be too stringent. Some of the pro-inflammantory genes that were deemed "significant" in the [@HDAC_2017] paper were filtered out with Bonferroni, so I used the Benjamini-Hockberg adjustment method instead.\

I used the epiriical Bayes method to compute differential expression, with trend=TRUE set to specify RNAseq data.

```{r create-DE-data-matrix}
#CREATE DATA MATRIX

model_design <- model.matrix(~ samples$treatment)
expressionMatrix <- as.matrix(final.df[,2:ncol(final.df)])
#USE LIMMA PACKAGE
minimalSet <- ExpressionSet(assayData=expressionMatrix)
fit <- lmFit(minimalSet, model_design)
#Apply empircal Bayes to compute differential expression
fit2 <- eBayes(fit,trend=TRUE)

adjust.method = "BH"
#adjust.method="bonferroni"
topfit <- topTable(fit2, 
                   sort.by="none",
                   coef=ncol(model_design),
                   adjust.method = adjust.method,
                   number = nrow(expressionMatrix))

topfit <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],topfit, stringsAsFactors = FALSE)
colnames(topfit)[1] <- "hgnc_symbol"


```


## Plot of differentially expressed genes

```{r DE-plot}
title <- "Figure 1: DE genes between DMSO and HDAC treated dHAECs."
plot(topfit$AveExpr,topfit$logFC, main=title, xlab= "Average expression", ylab= "log fold-change")
pval <- 0.05
# create up and down indices using pval cutoff and logFC
down.index <- which((topfit$adj.P.Val < pval) & (topfit$logFC < 0))
up.index <- which((topfit$adj.P.Val < pval) & (topfit$logFC > 0))
# colour points accordingly
points(topfit$AveExpr[up.index], topfit$logFC[up.index], col = "green")
points(topfit$AveExpr[down.index], topfit$logFC[down.index], col = "red")
#diabetic inflammatory genes investigated in paper
D.inflamm.genes <- c("VCAN", "CD58", "CD200",  "SIGIRR", "CCL2", "ICAM2")
D.inflamm.index <- which(topfit$hgnc_symbol %in% D.inflamm.genes )
points(topfit$AveExpr[D.inflamm.index], topfit$logFC[D.inflamm.index], col = "blue")
text(topfit$AveExpr[D.inflamm.index], topfit$logFC[D.inflamm.index], labels=D.inflamm.genes, cex= 0.7, pos=2, col = "blue")

num.genes <- 2
# Highest logFC genes
top.up.genes <- head(topfit[order(topfit$logFC, decreasing = TRUE),], num.genes)$hgnc_symbol
top.up.genes.idx <- which(topfit$hgnc_symbol %in% top.up.genes )
labels.up <- topfit$hgnc_symbol[top.up.genes.idx]
text(topfit$AveExpr[top.up.genes.idx], topfit$logFC[top.up.genes.idx], labels=labels.up, cex= 0.7, pos=2)
#Lowest logFC genes
top.down.genes <- head(topfit[order(topfit$logFC, decreasing = FALSE),], num.genes)$hgnc_symbol
top.down.genes.idx <- which(topfit$hgnc_symbol %in% top.down.genes )
labels.down <- topfit$hgnc_symbol[top.down.genes.idx]
text(topfit$AveExpr[top.down.genes.idx], topfit$logFC[top.down.genes.idx], labels=labels.down, cex= 0.7, pos=2)
# LEGEND
legend(10, 7,legend=c("Upregulated", "Downregulated", "Diabetic genes"),col=c("green", "red" ,"blue"), pch=1, cex=0.8)

```

## Counts of DE genes
I used a threshold of 0.05 cutoff for the adjusted p-value. My DE results returened a total of 6600/13175 genes deemed significantly differentially expressed. Of these, 3291 were significantly upregulated and 3309 significantly downregulated. I include labels for diabetic genes, as well as the top 2 upregulated and downregulated genes (Figure 1).


# Clustering of top hits using heatmap
The heatmap shows that DMSO and SAHA treated are clustering separately.

```{r heatmap}
#Make HEATMAP ==> TAKES LONG TIME, ONLY PRINT HEATMAP AT VERY END
# get significant indices
significant.index <- which((topfit$adj.P.Val < pval))
heatmap_matrix <- final.df[significant.index, 2:ncol(final.df)]
#heatmap_matrix <- head(heatmap_matrix, 10)

heatmap_matrix <- t(scale(t(heatmap_matrix)))
heatmap_col = colorRamp2(c(min(heatmap_matrix), 0, max(heatmap_matrix)), c("blue", "white", "red"))

current_heatmap <- Heatmap(as.matrix(heatmap_matrix),
                               column_title = "Figure 2: Clustering for top hits between SAHA and DMSO treated dHAECs",                              
                               show_row_dend = TRUE,
                               show_column_dend = TRUE,
                               col=heatmap_col,
                               show_column_names = TRUE,
                               show_row_names = FALSE,
                               show_heatmap_legend = TRUE
                               , 

heatmap_legend_param = list(
        title = "Legend", at = c(-2, 0, 2), labels = c("min", "zero", "max")
    )

)

current_heatmap

```


# Investigate data (using top hits)
```{r gprofiler-queries}
pval <- 0.05

#ALL
sign.hits <- topfit[which((topfit$adj.P.Val < pval)),]
#SORT BY DECREASING log_FC
sign.hits <- sign.hits[order(sign.hits$adj.P, decreasing = TRUE),]
#REMOVE ROWS WITH NO HGNC SYMBOL
sign.hits <- sign.hits[sign.hits$hgnc_symbol != "",]
# GENERATE RANK FILE FOR PRERANKED GSEA
rank.df <- sign.hits[c("hgnc_symbol","logFC")]
write.table(rank.df, file=paste0(getwd(), "/", "sign_hits.rnk"), quote=FALSE, sep='\t', row.names=FALSE)
```

# Thresholded analysis using GSEA

```{sh}
#:%s/Details ...\t//g
#:%s/GS DETAILS\t//g
# Need to remove "GS DETAILS" column from files gsea_report_for_na_neg_1584894635307.xls and gsea_report_for_na_pos_1584894635307.xls

```


# GSEA RESULTS

```{r}
#neg.file <- paste0(getwd(),"/GSEA_outdir/A3_GSEA.GseaPreranked.1584894635307/gsea_report_for_na_neg_1584894635307.xls")
neg.file <- "tmp.xls"
read.table(file=neg.file, header = TRUE, stringsAsFactors = F, sep="\t")

```




# R package references
GEOmetadb https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html \ 

edgeR https://bioconductor.org/packages/release/bioc/html/edgeR.html \ 

biomaRt https://bioconductor.org/packages/release/bioc/html/biomaRt.html \ 

tidyr https://cran.r-project.org/web/packages/tidyr/index.html \ 

dplyr https://cran.r-project.org/web/packages/dplyr/index.html \ 

ComplexHeatmap https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html \ 

circlize https://cran.r-project.org/web/packages/circlize/index.html \ 

gprofiler2 https://cran.r-project.org/web/packages/gprofiler2/index.html\

stringr https://cran.r-project.org/web/packages/stringr/stringr.pdf\

kableExtra https://cran.r-project.org/web/packages/kableExtra/index.html\

# Paper references


```{r, message=FALSE, echo=FALSE}
#SCRAP
# WNT
# Wnt <-gprofiler.results[grep("Wnt|wnt", gprofiler.results$term_name),]
# small.terms <- gprofiler.results[gprofiler.results$term_size < 200,]
# wnt <- grep("Wnt", small.terms$term_name)

#SCRAP
# ANALYSIS OF EXPRESSION OF WNT TARGET GENES
# wnt.targets <- c("MYC", "CCDN1", "VEGF", "MMP7", "LEF1", "TCF7L2")
# topfit[topfit$hgnc_symbol %in% wnt.targets,]
#gprofiler.results <- gprofiler2::gprofiler2(noquote(sign.hits$hgnc_symbol), organism = "hsapiens" )
# #gprofiler(query, organism = "hsapiens", sort_by_structure = T,
#   ordered_query = F, significant = T, exclude_iea = F,
#   underrep = F, evcodes = F, region_query = F, max_p_value = 1,
#   min_set_size = 0, max_set_size = 0, min_isect_size = 0,
#   correction_method = "analytical", hier_filtering = "none",
#   domain_size = "annotated", custom_bg = "", numeric_ns = "",
#   png_fn = NULL, include_graph = F, src_filter = NULL)

#ORDER BY DECREASING P-VALUE
#head(sign.hits)
# SORT BY DECREASING EXPRESSION VALUE
#sign.hits <- sign.hits[order(sign.hits.pval$logFC, decreasing = TRUE),]

# testing <- 0
# if (testing){
#     output_hits.pval[order(output_hits.pval$AveExpr),]
#     nrow(output_hits.pval)
#     # MAYBE COLOUR THESE DOTS IN THE MA PLOT (THE INCREASE/DECREASE EXPRESSION PLOT)
#     output_hits.pval[output_hits.pval$hgnc_symbol == "MAST1",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "VCAN",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "CD58",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "CD200",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "SIGIRR",]
# }
# #UPREGULATED
# output_hits.pval.up <- output_hits.pval[which(output_hits.pval$logFC > 0),]
# length(which(output_hits.pval.up$adj.P.Val < pval))
# up.down="up."
# #cat(noquote(output_hits.pval$hgnc_symbol[which(output_hits.pval$adj.P.Val < pval)]))
# #write(output_hits.pval$hgnc_symbol[which(output_hits.pval.up$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )
# 
# 
# output_hits.pval.down <- output_hits.pval[which(output_hits.pval$logFC < 0),]
# output_hits.pval.down <- output_hits.pval[order(output_hits.pval$logFC, decreasing = FALSE),]
# length(which(output_hits.pval.up$adj.P.Val < pval))
# up.down="down."
# # write(output_hits.pval$hgnc_symbol[which(output_hits.pval.down$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )


#BP
# pathways <- gprofiler.results[gprofiler.results$source =="GO:BP",]$term_name
# pathways.up <- gprofiler.results.up[gprofiler.results.up$source =="GO:BP",]$term_name
# pathways.down <- gprofiler.results.down[gprofiler.results.down$source =="GO:BP",]$term_name

# gprofiler.results.down$term_name[grep("Wnt|wnt", gprofiler.results.up$term_name)]$term_name
# #Wnt GO:BP only
# wnt.BP <- pathways[grep("Wnt|wnt", pathways)]
# wnt.BP.up <-pathways.up[grep("Wnt|wnt", pathways.up)]
# wnt.BP.down <- pathways.down[grep("Wnt|wnt", pathways.down)]
# 
# cols <- colnames(gprofiler.results)
# sources <-unique(gprofiler.results$source)
# gprofiler.results.pathways <- gprofiler.results[gprofiler.results$source =="GO:BP",]
# gprofiler.results.kegg <- gprofiler.results[gprofiler.results$source =="KEGG",]
# gprofiler.results.reactome <- gprofiler.results[gprofiler.results$source =="REAC",]

```


