---
title: "Assignment 2: Differential expression analysis and investigation of biological effects of SAHA treatment of diabetic human aortic endothelial cells"
output: 
  html_document:
    toc: true
    toc_depth: 2
bibliography: A3_michaelapostolides.bib
---

**Name:** Michael Apostolides\

```{r, message=FALSE, echo=FALSE}
# check to ensure all needed packages are installed
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

if (!requireNamespace("GEOmetadb", quietly = TRUE))
  BiocManager::install("GEOmetadb")

if (!requireNamespace("edgeR", quietly = TRUE))
  BiocManager::install("edgeR")

if (!requireNamespace("biomaRt", quietly = TRUE))
  BiocManager::install("biomaRt")

if (!requireNamespace("ComplexHeatmap", quietly = TRUE))
  BiocManager::install("ComplexHeatmap")

if (!requireNamespace("gProfileR", quietly = TRUE))
  BiocManager::install("gProfileR")

if (!requireNamespace("gprofiler2", quietly = TRUE))
  BiocManager::install("gprofiler2")

if (!requireNamespace("stringr", quietly = TRUE))
  install.packages("stringr")

if (!requireNamespace("kableExtra", quietly = TRUE))
  install.packages("kableExtra")
```


```{r, message=FALSE, echo=FALSE}
# load needed libraries
library("GEOmetadb")
library("edgeR")
library("biomaRt")
library("tidyr")
library("dplyr")
library("ComplexHeatmap")
library(circlize)
library("gprofiler2")
library("stringr")
library(kableExtra)
```


# Download expression data from GEO
```{r, message=FALSE, echo=FALSE}
#GET THE EXPRESSION DATA
sfiles = getGEOSuppFiles('GSE77108')
fnames = rownames(sfiles)
# there is only one supplemental file
HDAC = read.delim(fnames[1],header=TRUE, check.names = FALSE)
#SUBSET ONLY COLUMNS WHICH have diabetic cells and SAHA or DMSO treatment
SAHA.DMSO <- HDAC[grep("DMSO|SAHA|Geneid", colnames(HDAC))]
diabetic.SAHA.DMSO <- SAHA.DMSO[grep("D-|Geneid", colnames(SAHA.DMSO))]
```


```{r, echo=FALSE}
#GENERATE "samples", METADATA ABOUT SAMPLES
samples <- data.frame(lapply(colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)], 
                             FUN=function(x){unlist(strsplit(x, split = "\\."))[c(2)]}))
colnames(samples) <- colnames(diabetic.SAHA.DMSO)[2:ncol(diabetic.SAHA.DMSO)]
rownames(samples) <- c("treatment")
samples <- data.frame(t(samples), stringsAsFactors=FALSE)

#LABEL CELL TYPES AS "DMSO" TREATED OR "SAHA" TREATED
DMSO <- grep("DMSO", rownames(samples))
SAHA <- grep("SAHA", rownames(samples))
samples$treatment[DMSO] = "DMSO"
samples$treatment[SAHA] = "SAHA"
```


```{r, echo=FALSE, message=FALSE}
#FILTER OUT GENES THAT HAVE LOW COUNTS (CLEANING DATA)
#use edgeR to calculate counts per million (cpms)
cpms = cpm(diabetic.SAHA.DMSO[2:ncol(diabetic.SAHA.DMSO)])
rownames(cpms) <- diabetic.SAHA.DMSO[,1]
#use cpms to determine which of our ENSGs to throw out
keep = rowSums(cpms >1) >=3
diabetic.SAHA.DMSO.filtered = diabetic.SAHA.DMSO[keep,]
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$Geneid
```



```{r map_ensg_to_hgnc, echo=FALSE, message=FALSE, warning=FALSE}
#MAP ENSG TO HGNC SYMBOLS
#get mart object from useast mirror (this mirror was working at the time of running)
ensembl <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "uswest")#useast

conversion_stash <- "SAHA_id_conversion.rds"
if(file.exists(conversion_stash)){
  HDAC_id_conversion <- readRDS(conversion_stash)
} else {
  HDAC_id_conversion <- getBM(attributes = c("ensembl_gene_id","hgnc_symbol"),
                               filters = c("ensembl_gene_id"),
                               values = diabetic.SAHA.DMSO.filtered$Geneid,
                               mart = ensembl)
  saveRDS(HDAC_id_conversion, conversion_stash)
}


#rename col in diabetic.SAHA.DMSO.filtered  to match HDAC_id_conversion
colnames(diabetic.SAHA.DMSO.filtered)[colnames(diabetic.SAHA.DMSO.filtered) == "Geneid"] <- "ensembl_gene_id"
#join by ensembl_gene_id
diabetic.SAHA.DMSO.filtered <- dplyr::inner_join(diabetic.SAHA.DMSO.filtered , HDAC_id_conversion, by="ensembl_gene_id" )
#make row names ensembl_gene_id
rownames(diabetic.SAHA.DMSO.filtered) <- diabetic.SAHA.DMSO.filtered$ensembl_gene_id
#remove ensembl_gene_id column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[colnames(diabetic.SAHA.DMSO.filtered) != "ensembl_gene_id"]
#make hgnc_symbol first column
diabetic.SAHA.DMSO.filtered <- diabetic.SAHA.DMSO.filtered[c(7,1:6)]
#stop("Stop")

```





# INTRODUCTION  
The dataset is from [GEO GSE77108](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77108)[@HDAC_2017], and selects one of several treatment analysis in this paper. This paper corresponds to the GEO dataset with ID GSE77108. The treatment selected is SAHA treatment of human aortic endothelial cells obtained from a diabetic individual (dHAECs) line as compared to control DMSO treatment of those same dHAECs. My workflow begins with the RNAseq read counts provided by the authors, which were obtained using the "Counts" feature of the STAR RNAseq aligner[@HDAC_2017].   
Normalization has already been done by the authors using the HPRT1 housekeeping gene[@HDAC_2017]. However, I have gone ahead and normalized the data again anyway using TMM:

## TMM normalization
```{r TMM-normalization}
#APPLYING TMM NORMALIZATION TO THE DATASET
filtered_data_matrix <- as.matrix(diabetic.SAHA.DMSO.filtered[,2:ncol(diabetic.SAHA.DMSO)])
rownames(filtered_data_matrix) <- rownames(diabetic.SAHA.DMSO.filtered)
d = DGEList(counts=filtered_data_matrix, group=samples$treatment)
d = calcNormFactors(d)

final.df <- cpm(d)
final.df.log2 <- log2(cpm(d))
# add hgnc symbols as first column
final.df.log2 <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],final.df.log2, stringsAsFactors = FALSE)
colnames(final.df.log2)[1] <- "hgnc_symbol"

# The reason we were getting -Inf values is because the values were "0" prior to log2 transformation
#remove values which have -Inf as values in matrix
final.df <- final.df.log2
final.df[final.df== -Inf] <- 0

```


# Differential Gene Expression 

I used the Benjamini-Hockberg adjustment method. I initially tried the Bonferroni method, but it proved to be too stringent. Some of the pro-inflammantory genes that were deemed "significant" in the [@HDAC_2017] paper were filtered out with Bonferroni, so I used the Benjamini-Hockberg adjustment method instead.\

I used the epiriical Bayes method to compute differential expression, with trend=TRUE set to specify RNAseq data.

```{r create-DE-data-matrix}
#CREATE DATA MATRIX

model_design <- model.matrix(~ samples$treatment)
expressionMatrix <- as.matrix(final.df[,2:ncol(final.df)])
#USE LIMMA PACKAGE
minimalSet <- ExpressionSet(assayData=expressionMatrix)
fit <- lmFit(minimalSet, model_design)
#Apply empircal Bayes to compute differential expression
fit2 <- eBayes(fit,trend=TRUE)

adjust.method = "BH"
#adjust.method="bonferroni"
topfit <- topTable(fit2, 
                   sort.by="none",
                   coef=ncol(model_design),
                   adjust.method = adjust.method,
                   number = nrow(expressionMatrix))

topfit <-cbind.data.frame(diabetic.SAHA.DMSO.filtered[,1],topfit, stringsAsFactors = FALSE)
colnames(topfit)[1] <- "hgnc_symbol"


```



```{r}

#diabetic inflammatory genes investigated in paper
D.inflamm.genes <- c("VCAN", "CD58", "CD200",  "SIGIRR", "CCL2", "ICAM2")


```



# Investigate data (using top hits)
```{r gprofiler-queries}
# #ALL
# sign.hits <- topfit[which((topfit$adj.P.Val < pval)),]
# #SORT BY DECREASING log_FC
# sign.hits <- sign.hits[order(sign.hits$adj.P, decreasing = TRUE),]

# # GENERATE RANK FILE FOR PRERANKED GSEA
# rank.df <- sign.hits[c("hgnc_symbol","logFC")]


#compute ranks
ranks <- -log(topfit$adj.P.Val, base=10)*sign(topfit$logFC)
genes <- topfit$hgnc_symbol
rnk.df <- data.frame(GeneName=genes, rank=ranks, stringsAsFactors = FALSE)
#sort table by ranks
rnk.df <- rnk.df[order(rnk.df$rank, decreasing = TRUE),]
rows <- nrow(rnk.df)
#REMOVE ROWS WITH NO HGNC SYMBOL
rnk.df <- rnk.df[rnk.df$GeneName != "",]
rows.no_hgnc <- as.character(rows - nrow(rnk.df))
print(paste0("Number of entries without an hgnc symbol: ", rows.no_hgnc))
write.table(rnk.df, file=paste0(getwd(), "/", "sign_hits.unthresholded.rnk"), quote=FALSE, sep='\t', row.names=FALSE)

#write expression values file
expression_values <- data.frame(topfit$hgnc_symbol, topfit$logFC)
write.table(expression_values,  file=paste0(getwd(), "/", "expression_values.tsv"), sep='\t', quote=FALSE, row.names=FALSE )

```

# Non-Thresholded analysis using GSEA

I used GSEA (**citation for GSEA**) for non-thresholded analysis. I used the same geneset file which was used in the class homework:\ [[http://download.baderlab.org/EM_Genesets/February_01_2020/Human/symbol/Human_GOBP_AllPathways_no_GO_iea_February_01_2020_symbol.gmt]]\

## Summary of enrichment results
```{r}
#file:///Users/andreeduquette/BCB420/student_apostonaut/assignment3/GSEA_outdir/A3_GSEA_UNTHRESHOLDD.GseaPreranked.1585427734453/gsea_report_for_na_pos_1585427734453.xls
neg.file <- paste0(getwd(),"/GSEA_outdir/A3_GSEA_UNTHRESHOLDD.GseaPreranked.1585427734453/gsea_report_for_na_neg_1585427734453.xls")
neg.file <- read.table(file=neg.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")
pos.file <- paste0(getwd(),"/GSEA_outdir/A3_GSEA_UNTHRESHOLDD.GseaPreranked.1585427734453/gsea_report_for_na_pos_1585427734453.xls")
pos.file <- read.table(file=pos.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")

top.pos <- head(pos.file$NAME, 15)
top.neg <- head(neg.file$NAME, 15)

df.terms <- data.frame(Upregulated=top.pos, Downregulated=top.neg)
title <- "Table: Comparison of top 15 GSEA upregulated and downregulated gene sets"
df.terms  %>%
  kable(caption = title) %>%
  kable_styling()
```
```{sh}
#:%s/Details ...\t//g
#:%s/GS DETAILS\t//g
#%s/\t\t/\t/g
# Need to remove "GS DETAILS" column from files gsea_report_for_na_neg_1584894635307.xls and gsea_report_for_na_pos_1584894635307.xls

```

I have included a table of the top 15 upregulated and downregulated terms of the GSEA results (Table 1). A summary of the contents for the top hits is as follows:\
* Upregulated: the main themes include: "cognition/learning/memory", "Synaptic signalling", "ion transport/channels", "lipid/lipoprotein" (Table 1).
* Downregulated: the main themes include: "E2F", "DNA replication", "ATR signaling/replication stress", "cell cycle checkpoints" (Table 1).

<!-- How do these results compare to the results from the thresholded analysis in Assignment #2. Compare qualitatively. Is this a straight forward comparison? Why or why not? -->
## Comparison of GSEA and g:profiler results
Upregulated: protein localization, lytic vacuole/lysosome, Axonogenesis, wnt signaling, (Ras) GTPase binding, and synapse-related terms.
Downregulated: DNA replication, cell cycle checkpoint, and chromosome segregation gene sets, retinoblastoma Gene in Cancer 

* Upregulated comparison: the upregulated terms have only synapse-related terms in common. There are major differences in the top 15 hits
* Downregulated comparison: the downregulated terms have "DNA replication" and "cell cycle checkpoint" terms in common. 
Comparing the outputs directly using the tables provided is difficult. There are many keywords, and finding themes among terms is time-consuming to do manually.

# Network diagram and HDAC inhibitors

<Network Diagram>


# Wnt

I have chosen the Wnt pathway as a pathway to investigate in the GSEA non-thresholded results. Wnt-related gene sets are make up 11/296 significant gene sets with p-value threshold of 0.01, including ranks #23 and #44 in "na_pos" GSEA results. There are no Wnt-related terms in the "na_neg" downregulated sets, suggesting that Wnt-related effects are uniquely seen in SAHA-treated dHAECs. There is also literature support for modulation of the Wnt pathway by SAHA.  In the (Götze et al. 2014) paper, they show that SAHA attenuates proliferation and induces apoptosis in Wnt-activated colorectal carcinoma cell lines. In the (Sujeong Jang 2018b) paper, they show that expression of most of the Wnt-related genes they were investigating were highly increased following treatment with HDAC inhibitors.

The Wnt pathway is a well-defined pathway with literature support for SAHA effects, so it makes a good candidate for more detailed investigation. 

```{r}
pval <- 0.01
pos.file.sign <- pos.file[pos.file$NOM.p.val < pval,]
wnt.pos.sign <- pos.file.sign[grep("WNT", pos.file.sign$NAME),]
print(nrow(wnt.pos.sign))


neg.file.sign <- neg.file[neg.file$NOM.p.val < pval,]
wnt.neg.sign <- neg.file.sign[grep("WNT", neg.file.sign$NAME),]

leading.edge.genes.file <- paste0(getwd(), "/GSEA_outdir/A3_GSEA_UNTHRESHOLDD.GseaPreranked.1585427734453/NEGATIVE\ REGULATION\ OF\ CANONICAL\ WNT\ SIGNALING\ PATHWAY%GOBP%GO:0090090.xls")
leading.edge.genes.file <- read.table(file=leading.edge.genes.file, header = T, stringsAsFactors = F, sep="\t", quote = "\"")
leading.edge <- leading.edge.genes.file[leading.edge.genes.file$CORE.ENRICHMENT == "Yes",]
leading.edge <- data.frame(PROBE=leading.edge$PROBE, RANK.METRIC.SCORE=leading.edge$RANK.METRIC.SCORE)
leading.edge.genes <- leading.edge$PROB
colnames(leading.edge.genes.file)

wikipathway.genes <- paste0(getwd(), "/", "Wnt.pathway.WikiPathways.txt")
wikipathway.genes <- read.table(file=wikipathway.genes, header = F, stringsAsFactors = F, sep="\t", quote = "\"")
wikipathway.genes <- wikipathway.genes$V1
noquote(intersect(wikipathway.genes, leading.edge.genes))
```


# Annotated Wnt pathway diagram from WikiPathways

I have downloaded the WP428 (Wnt Signaling) WikiPathways network using the "Import Network from Public Databases" feature of Cytoscape, and annotated it with the rank file I used as input to my GSEA run (Figure 2). Additionally, the GO term "GO:0090090", "negative regulation of canonical Wnt signaling pathway", which is hit #23 of GSEA's "na_pos" results, has the following genes in common with WikiPathwasy WP428: CSNK1A1, DKK1, DKK2, DVL2, GSK3B, NOTUM, SFRP1, SOX17. All of these genes are in the leading edge of this GSEA result.
Additionally, there are 44 other genes in the leading edge which are not listed in Figure 2, but which are known to negatively regulated Wnt signalling.

DKK1, SFRP1 are both highly ranked Wnt pathway inhibitors, as can be seen by the (----| ) inhibitory arrows in Figure 2. SFRP1 inhibits signaling of Wnt proteins directly, whereas  DKK1 inhibits LRP5/6, which in turn interact with Frizzled(FZD) receptors (Figure 2). DKK1, a known Wnt antagonist, is epigenetically silenced in a medulloblastoma cell line, and is upregulated upon HDAC inhibition [@DKK1_HDAC]. Epigenetic silencing of SFRP1 leads to increased survival of renal and breast cancers, and re-rexpression of SFRP1 upon treatment with HDAC inhibitors coincides with reduced cell growth and apoptosis in chemoresistant cancers [@SFRP1_HDAC]. 

<IMAGE>
```{r}
wnt.wikipathways.image <- paste0(getwd(), "/", "Wnt-WikiPathways-Annotated.png")
include_graphics(img1_path)
```

![Caption for the picture.]("/files/Wnt-WikiPathways-Annotated.png")

# R package references
GEOmetadb https://www.bioconductor.org/packages/release/bioc/html/GEOmetadb.html \ 

edgeR https://bioconductor.org/packages/release/bioc/html/edgeR.html \ 

biomaRt https://bioconductor.org/packages/release/bioc/html/biomaRt.html \ 

tidyr https://cran.r-project.org/web/packages/tidyr/index.html \ 

dplyr https://cran.r-project.org/web/packages/dplyr/index.html \ 

ComplexHeatmap https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html \ 

circlize https://cran.r-project.org/web/packages/circlize/index.html \ 

gprofiler2 https://cran.r-project.org/web/packages/gprofiler2/index.html\

stringr https://cran.r-project.org/web/packages/stringr/stringr.pdf\

kableExtra https://cran.r-project.org/web/packages/kableExtra/index.html\

# Paper references


```{r, message=FALSE, echo=FALSE}
#SCRAP
# WNT
# Wnt <-gprofiler.results[grep("Wnt|wnt", gprofiler.results$term_name),]
# small.terms <- gprofiler.results[gprofiler.results$term_size < 200,]
# wnt <- grep("Wnt", small.terms$term_name)

#SCRAP
# ANALYSIS OF EXPRESSION OF WNT TARGET GENES
# wnt.targets <- c("MYC", "CCDN1", "VEGF", "MMP7", "LEF1", "TCF7L2")
# topfit[topfit$hgnc_symbol %in% wnt.targets,]
#gprofiler.results <- gprofiler2::gprofiler2(noquote(sign.hits$hgnc_symbol), organism = "hsapiens" )
# #gprofiler(query, organism = "hsapiens", sort_by_structure = T,
#   ordered_query = F, significant = T, exclude_iea = F,
#   underrep = F, evcodes = F, region_query = F, max_p_value = 1,
#   min_set_size = 0, max_set_size = 0, min_isect_size = 0,
#   correction_method = "analytical", hier_filtering = "none",
#   domain_size = "annotated", custom_bg = "", numeric_ns = "",
#   png_fn = NULL, include_graph = F, src_filter = NULL)

#ORDER BY DECREASING P-VALUE
#head(sign.hits)
# SORT BY DECREASING EXPRESSION VALUE
#sign.hits <- sign.hits[order(sign.hits.pval$logFC, decreasing = TRUE),]

# testing <- 0
# if (testing){
#     output_hits.pval[order(output_hits.pval$AveExpr),]
#     nrow(output_hits.pval)
#     # MAYBE COLOUR THESE DOTS IN THE MA PLOT (THE INCREASE/DECREASE EXPRESSION PLOT)
#     output_hits.pval[output_hits.pval$hgnc_symbol == "MAST1",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "VCAN",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "CD58",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "CD200",]
#     output_hits.pval[output_hits.pval$hgnc_symbol == "SIGIRR",]
# }
# #UPREGULATED
# output_hits.pval.up <- output_hits.pval[which(output_hits.pval$logFC > 0),]
# length(which(output_hits.pval.up$adj.P.Val < pval))
# up.down="up."
# #cat(noquote(output_hits.pval$hgnc_symbol[which(output_hits.pval$adj.P.Val < pval)]))
# #write(output_hits.pval$hgnc_symbol[which(output_hits.pval.up$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )
# 
# 
# output_hits.pval.down <- output_hits.pval[which(output_hits.pval$logFC < 0),]
# output_hits.pval.down <- output_hits.pval[order(output_hits.pval$logFC, decreasing = FALSE),]
# length(which(output_hits.pval.up$adj.P.Val < pval))
# up.down="down."
# # write(output_hits.pval$hgnc_symbol[which(output_hits.pval.down$adj.P.Val < pval)], file = paste0("gene_hits","/",adjust.method,".",pval,".",up.down,"gene_hits.txt") )


#BP
# pathways <- gprofiler.results[gprofiler.results$source =="GO:BP",]$term_name
# pathways.up <- gprofiler.results.up[gprofiler.results.up$source =="GO:BP",]$term_name
# pathways.down <- gprofiler.results.down[gprofiler.results.down$source =="GO:BP",]$term_name

# gprofiler.results.down$term_name[grep("Wnt|wnt", gprofiler.results.up$term_name)]$term_name
# #Wnt GO:BP only
# wnt.BP <- pathways[grep("Wnt|wnt", pathways)]
# wnt.BP.up <-pathways.up[grep("Wnt|wnt", pathways.up)]
# wnt.BP.down <- pathways.down[grep("Wnt|wnt", pathways.down)]
# 
# cols <- colnames(gprofiler.results)
# sources <-unique(gprofiler.results$source)
# gprofiler.results.pathways <- gprofiler.results[gprofiler.results$source =="GO:BP",]
# gprofiler.results.kegg <- gprofiler.results[gprofiler.results$source =="KEGG",]
# gprofiler.results.reactome <- gprofiler.results[gprofiler.results$source =="REAC",]

```


